File: survey/static/survey/js/branching.js
// Progressive enhancement for conditional questions (skip logic).
//
// The server is authoritative: it decides which questions render and never
// requires/stores a hidden question on submit. This script only improves the
// SAME-PAGE case — revealing or hiding a dependent live as its controlling
// answer changes — so studies work without JS (cross-page) and feel
// responsive with it (same-page).
//
// A conditional question's <fieldset> carries data-visible-if='<rule JSON>'.
// We only manage a fieldset whose controlling questions are present ON THIS
// page; if a controller lives on an earlier page (not in this form), we leave
// the field visible because the server already decided to render it.
//
// The operator semantics below MUST mirror experiments/branching.py
// (_eval_clause) — answers read from the DOM are strings, so both engines
// compare "loosely": numbers numerically when both sides parse, strings
// otherwise. Change the two files together.

const fieldName = (qid) => `q_${qid}`;

// Read the current answer for a question from this form's inputs. Inputs
// disabled by this script (hidden dependents) don't submit, so they read as
// unanswered — keeping chained rules in sync with what the server will see.
// Returns undefined when the control is not on this page at all.
const readAnswer = (form, qid) => {
  const radios = [...form.querySelectorAll(`input[type=radio][name="${fieldName(qid)}"]`)];
  if (radios.length) {
    return radios.find((r) => r.checked && !r.matches(":disabled"))?.value ?? null;
  }
  const checks = [...form.querySelectorAll(`input[type=checkbox][name="${fieldName(qid)}"]`)];
  if (checks.length) {
    return checks.filter((c) => c.checked && !c.matches(":disabled")).map((c) => c.value);
  }
  const el = form.querySelector(`[name="${fieldName(qid)}"]`);
  if (el) return el.matches(":disabled") || el.value === "" ? null : el.value;
  return undefined;
};

// The numeric-literal grammar BOTH engines share: plain decimal / scientific
// notation only (Python's float() additionally accepts "inf"/"1_0", JS's
// Number() accepts "0x10" — the shared regex removes the drift).
const NUMERIC_RE = /^[+-]?(?:\d+\.?\d*|\.\d+)(?:[eE][+-]?\d+)?$/;

const toNumber = (x) => {
  if (typeof x === "number") return x;
  if (typeof x === "string" && NUMERIC_RE.test(x.trim())) return Number(x);
  return null;
};

// Type-tolerant equality: a rating answered "3" in the DOM matches a rule
// value authored as the number 3; lists compare element-wise loose (mirrors
// _loose_eq in branching.py — change the two together).
const looseEq = (a, b) => {
  if (Array.isArray(a) && Array.isArray(b)) {
    return a.length === b.length && a.every((item, i) => looseEq(item, b[i]));
  }
  if (Array.isArray(a) || Array.isArray(b)) return false;
  if (a == null || b == null) return a === b;
  const na = toNumber(a);
  const nb = toNumber(b);
  if (na !== null && nb !== null) return na === nb;
  return String(a) === String(b);
};

// "in"/"nin" membership: list values test loose element equality; a string
// value tests substring membership of a SCALAR answer (an array answer is
// never "in" a string — String([]) === "" would match everything); anything
// else is an empty collection.
const membership = (ans, value) => {
  if (Array.isArray(value)) return value.some((item) => looseEq(ans, item));
  if (typeof value === "string") return !Array.isArray(ans) && value.includes(String(ans));
  return false;
};

const isAnswered = (ans) =>
  ans != null && ans !== "" && !(Array.isArray(ans) && ans.length === 0);

const clauses = (rule) => rule.all ?? rule.any ?? [rule];

const refsOnPage = (rule, form) =>
  clauses(rule).every((c) => readAnswer(form, c.question) !== undefined);

const clauseTrue = (c, form) => {
  const ans = readAnswer(form, c.question);
  if (c.op === "answered") return isAnswered(ans);
  if (c.op === "not_answered") return !isAnswered(ans);
  if (ans == null) return false;
  const v = c.value;
  switch (c.op) {
    case "eq":
      return looseEq(ans, v);
    case "ne":
      return !looseEq(ans, v);
    case "in":
      return membership(ans, v);
    case "nin":
      return !membership(ans, v);
    case "contains":
      return Array.isArray(ans)
        ? ans.some((item) => looseEq(item, v))
        : String(ans).includes(String(v));
    case "gt":
    case "lt":
    case "gte":
    case "lte": {
      const a = toNumber(ans);
      const b = toNumber(v);
      if (a === null || b === null) return false;
      return { gt: a > b, lt: a < b, gte: a >= b, lte: a <= b }[c.op];
    }
  }
  return false;
};

const visible = (rule, form) => {
  if (rule.all) return rule.all.every((c) => clauseTrue(c, form));
  if (rule.any) return rule.any.some((c) => clauseTrue(c, form));
  return clauseTrue(rule, form);
};

document.querySelectorAll("form").forEach((form) => {
  const conditional = form.querySelectorAll("[data-visible-if]");
  if (!conditional.length) return;

  const setShown = (el, shown) => {
    el.hidden = !shown;
    // Latent dependents are server-rendered as <fieldset hidden disabled>;
    // clear the fieldset-level disabled too or the inner inputs stay dead.
    if ("disabled" in el) el.disabled = !shown;
    el.querySelectorAll("input, select, textarea").forEach((input) => {
      input.disabled = !shown;
    });
  };

  const apply = () => {
    conditional.forEach((el) => {
      let rule;
      try {
        rule = JSON.parse(el.dataset.visibleIf);
      } catch {
        return;
      }
      if (refsOnPage(rule, form)) {
        setShown(el, visible(rule, form));
        return;
      }
      // Controllers we can't read. For a normally-rendered fieldset that
      // means an earlier page — trust the server's render (visible). For a
      // server-latent fieldset (data-latent) the controller is on THIS page
      // but unreadable (matrix/ranking/plugin inputs): keep it hidden — the
      // server reveals it on the POST re-render when its rule passes.
      setShown(el, el.dataset.latent !== "1");
    });
  };

  form.addEventListener("change", apply);
  form.addEventListener("input", apply);
  apply();
});

File: experiments/branching.py
"""Conditional display ("skip logic") for questions.

A :class:`~experiments.models.Question` may carry a ``visible_if`` rule that
hides it unless answers to *earlier* questions in the same section match. Rules
are evaluated server-side against the answers already collected for the current
stimulus (per-stimulus for stimulus questions) or for the demographic section,
so branching is deterministic and works without JavaScript when the controlling
question is on an earlier page. A small progressive-enhancement script
(``survey/js/branching.js``) reveals same-page dependents live.

Rule shapes (all JSON):

* ``{"question": <id>, "op": "eq", "value": "Yes"}`` — a single clause,
* ``{"all": [clause, ...]}`` — every clause must be true,
* ``{"any": [clause, ...]}`` — at least one clause must be true.

Operators: ``eq, ne, in, nin, gt, lt, gte, lte, contains, answered,
not_answered``. This module is pure (no Django/model imports) so both
``experiments.models`` validation and ``survey`` flow code can use it.
"""
from __future__ import annotations

import re
from typing import Any, Iterable

OPERATORS = {
    "eq",
    "ne",
    "in",
    "nin",
    "gt",
    "lt",
    "gte",
    "lte",
    "contains",
    "answered",
    "not_answered",
}

# Operators that do not take a ``value``.
VALUELESS_OPS = {"answered", "not_answered"}


def _is_answered(value: Any) -> bool:
    return value is not None and value != "" and value != [] and value != {}


# The numeric-literal grammar BOTH engines share: plain decimal / scientific
# notation only. Python's float() additionally accepts "inf"/"nan"/"1_0" and
# JS's Number() accepts "0x10" — restricting both to this regex keeps their
# comparisons identical by construction.
_NUMERIC_RE = re.compile(r"^[+-]?(?:\d+\.?\d*|\.\d+)(?:[eE][+-]?\d+)?$")


def _to_number(x: Any) -> float | None:
    """Parse ``x`` under the shared numeric grammar (None = not numeric)."""
    if isinstance(x, bool):
        return None
    if isinstance(x, (int, float)):
        return float(x)
    if isinstance(x, str) and _NUMERIC_RE.match(x.strip()):
        return float(x)
    return None


def _to_str(x: Any) -> str:
    # Mirror JS String(): booleans stringify lowercase.
    if isinstance(x, bool):
        return "true" if x else "false"
    return str(x)


def _loose_eq(a: Any, b: Any) -> bool:
    """Type-tolerant equality shared with the client-side mirror.

    A rating stored as the int ``3`` matches a rule value authored as ``"3"``
    (and vice versa): numbers compare numerically whenever both sides parse
    under the shared grammar, everything else falls back to string equality.
    Lists compare element-wise loose (checkbox answers are strings in the DOM
    but may be typed in an authored rule). The same semantics live in
    ``survey/js/branching.js`` (looseEq) — change the two together.
    """
    if isinstance(a, list) and isinstance(b, list):
        return len(a) == len(b) and all(_loose_eq(x, y) for x, y in zip(a, b))
    if isinstance(a, (list, dict)) or isinstance(b, (list, dict)):
        return a == b
    if a is None or b is None:
        return a == b
    na, nb = _to_number(a), _to_number(b)
    if na is not None and nb is not None:
        return na == nb
    return _to_str(a) == _to_str(b)


def _membership(ans: Any, target: Any) -> bool:
    """``in``/``nin`` membership: lists test loose element equality, a string
    target tests substring membership of a *scalar* answer (an array answer is
    never "in" a string — mirrors the JS guard), anything else is an empty
    collection."""
    if isinstance(target, list):
        return any(_loose_eq(ans, item) for item in target)
    if isinstance(target, str):
        if isinstance(ans, (list, dict)):
            return False
        return _to_str(ans) in target
    return False


def _eval_clause(clause: dict, answers: dict[int, Any]) -> bool:
    qid = clause.get("question")
    op = clause.get("op")
    target = clause.get("value")
    ans = answers.get(qid)
    present = qid in answers

    if op == "answered":
        return present and _is_answered(ans)
    if op == "not_answered":
        return not (present and _is_answered(ans))
    if not present:
        return False

    try:
        if op == "eq":
            return _loose_eq(ans, target)
        if op == "ne":
            return not _loose_eq(ans, target)
        if op == "in":
            return _membership(ans, target)
        if op == "nin":
            return not _membership(ans, target)
        if op == "contains":
            if isinstance(ans, list):
                return any(_loose_eq(item, target) for item in ans)
            if isinstance(ans, dict):
                return target in ans
            # Scalar answers coerce to string on both sides, so a rating
            # stored as int 35 still "contains" the value "5".
            return _to_str(target) in _to_str(ans)
        if op in {"gt", "lt", "gte", "lte"}:
            a, b = _to_number(ans), _to_number(target)
            if a is None or b is None:
                return False
            return {
                "gt": a > b,
                "lt": a < b,
                "gte": a >= b,
                "lte": a <= b,
            }[op]
    except (TypeError, ValueError):
        return False
    return False


def iter_clauses(condition: dict) -> Iterable[dict]:
    """Yield the individual clauses of a (possibly compound) condition."""
    if not condition:
        return
    if "all" in condition:
        yield from (condition.get("all") or [])
    elif "any" in condition:
        yield from (condition.get("any") or [])
    else:
        yield condition


def referenced_question_ids(condition: dict) -> list[int]:
    return [
        c.get("question")
        for c in iter_clauses(condition)
        if isinstance(c, dict) and isinstance(c.get("question"), int)
    ]


def evaluate_condition(condition: dict, answers: dict[int, Any]) -> bool:
    if not condition:
        return True
    if "all" in condition:
        return all(_eval_clause(c, answers) for c in (condition.get("all") or []))
    if "any" in condition:
        return any(_eval_clause(c, answers) for c in (condition.get("any") or []))
    return _eval_clause(condition, answers)


def is_visible(question, answers: dict[int, Any]) -> bool:
    """True if ``question`` should be shown given ``answers`` (qid → value)."""
    cond = getattr(question, "visible_if", None)
    if not cond:
        return True
    return evaluate_condition(cond, answers)
